contracts/protocol/pool/L2Pool.sol

```diff
@@ -0,0 +1,135 @@
+pragma solidity ^0.8.10;
+
+import {Pool} from './Pool.sol';
+import {IPoolAddressesProvider} from '../../interfaces/IPoolAddressesProvider.sol';
+import {IL2Pool} from '../../interfaces/IL2Pool.sol';
+import {CalldataLogic} from '../libraries/logic/CalldataLogic.sol';
+
+/**
+ * @title L2Pool
+ * @author Aave
+ * @notice Calldata optimized extension of the Pool contract allowing users to pass compact calldata representation
+ * to reduce transaction costs on rollups.
+ */
+contract L2Pool is Pool, IL2Pool {
+  /**
+   * @dev Constructor.
+   * @param provider The address of the PoolAddressesProvider contract
+   */
+  constructor(IPoolAddressesProvider provider) Pool(provider) {
+    // Intentionally left blank
+  }
+
+  /// @inheritdoc IL2Pool
+  function supply(bytes32 args) external override {
+    (address asset, uint256 amount, uint16 referralCode) = CalldataLogic.decodeSupplyParams(
+      _reservesList,
+      args
+    );
+
+    supply(asset, amount, msg.sender, referralCode);
```

Comment:
The assumption here is that if a smart contract wants to do actions on behalf of the user, they should instead uncompress calldata themselves and call the public function from the base class?
If the contract size limit allows, it could be interesting to also provide an optional entrypoint for contracts that don't unpack calldata so they can rely on your unpacking and don't have the added complexity in their own code
```solidity
function supply(bytes32 args, address onBehalfOf) external
```

Reply:
Yes, our current assumption is that if a smart contract wants to do actions on behalf of the user, they will handle calldata encoding etc themselves and just use the standard Pool functions afterwards.